processor/src/advice/providers.rs

```diff
@@ -443,3 +431,48 @@ impl AdviceProvider for RecAdviceProvider {
         self.provider.advance_clock()
     }
 }
+
+impl RecAdviceProvider {
+    // FINALIZATION
+    // --------------------------------------------------------------------------------------------
+    /// Consumes the advice provider and returns an ([AdviceInputs], [MemAdviceProvider]) tuple.
+    /// The [AdviceInputs] can be used to re-execute the program. The [MemAdviceProvider] can be
+    /// used to extract execution results.
+    ///
+    /// The returned [AdviceInputs] instance will contain only the non-deterministic inputs which
+    /// were requested during program execution.
+    pub fn finalize(self) -> (AdviceInputs, MemAdviceProvider) {
```

Comment:
Instead of returning a `(AdviceInputs, MemAdviceProvider)` tuple we could return `(AdviceInputs, (Vec<Felt>, SimpleAdviceMap, MerkleStore)`.